Updated IsolatedManager to take a callback that captures the remote command

jbradberry opened the pull request:
##### SUMMARY
It was discovered that running tasks against isolated nodes was failing to capture and store the command, working directory, and environment of the remote command being run by ansible-runner.  This PR and the associated ansible-runner one fixes this issue by creating a new artifact file with the missing information, and then when it gets synced back up to the controlling instance, write the values back into the relevant model.

ref: ansible/ansible-runner#244


##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
awx: 4.0.0
```

jbradberry committed:
Updated IsolatedManager to take a callback that captures the remote command

ryanpetrello reviewed (approved):
I like this 👍

ryanpetrello commented on awx/main/isolated/manager.py:
```diff
@@ -187,6 +188,13 @@ def check(self, interval=None):
+                if os.path.exists(command_path):
+                    with open(command_path, 'r') as f:
+                        data = json.load(f)
+                    self.check_callback(data['command'], data['cwd'], self.runner_params['envvars'].copy())
```
Okay, so `runner_params` isn't quite what we want here.  This is the environment variables the *controller* uses when it runs a playbook to sync data *to* the isolated node.

jbradberry committed:
Change the artifact to also capture the actual envvars data

jbradberry committed:
Do the env vars redaction for isolated nodes on this side

chrismeyersfsu commented on awx/main/tasks.py:
```diff
@@ -1212,7 +1221,8 @@ def run(self, pk, **kwargs):
                 isolated_manager_instance = isolated_manager.IsolatedManager(
-                    cancelled_callback=lambda: self.update_model(self.instance.pk).cancel_flag
+                    cancelled_callback=lambda: self.update_model(self.instance.pk).cancel_flag,
+                    check_callback=self.check_callback,
```
Think this should be `check_callback=self.check_handler`

jbradberry replied:
Right you are.  Fixed.

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -187,6 +188,13 @@ def check(self, interval=None):
+                if os.path.exists(config_path):
+                    with open(config_path, 'r') as f:
+                        data = json.load(f)
+                    self.check_callback(data)
```
Are we going to end up calling this multiple times?

jbradberry replied:
The way I'm doing it now, probably.  I suppose the only path out other than completing the loop is if an exception is raised, so it can come out of the loop without much, if any, loss.

jbradberry replied:
I've brought it out of the loop.

chrismeyersfsu commented on awx/main/tasks.py:
```diff
@@ -1077,6 +1077,15 @@ def status_handler(self, status_data, runner_config):
+        self.instance = self.update_model(self.instance.pk,
+                                          job_args=json.dumps(config['command']),
+                                          job_cwd=config['cwd'],
+                                          job_env=build_safe_env(config['env']))
```
`                    credential.credential_type.inject_credential(` <-- takes in and may modify self.safe_envs. We should repeat the logic performed here https://github.com/ansible/awx/blob/3f73176ef2ae0d952b02e579974c60b807ec172b/awx/main/tasks.py#L1074

_Ideally_ we would alter the code so we can converge the code for isolated and non-isolated.

jbradberry replied:
Is that why we did it this way?  I had wondered about that.

However, do we want to still do it in this case?  Somebody had suggested to me that the env might wind up with some different values on isolated nodes, which is why I decided to serialize it into an artifact file for transport back to the controlling node.

chrismeyersfsu replied:
`which is why I decided to serialize it into an artifact file for transport back to the controlling node` that part is ideal and correct. But those artifacts may contain sensitive vars so we need to redact it.

jbradberry replied:
That's what the `build_safe_env` call there is doing, if I understand things correctly.

chrismeyersfsu replied:
Yes, but that is one piece of the puzzle. The `safe_env` dict returned by `build_safe_env` is added to by `credential.credential_type.inject_credential()`. So we can't just call `build_safe_env`. We need to use the dict that has been added to by `credential.credential_type.inject_credential()`

jbradberry replied:
:face_with_head_bandage:

jbradberry replied:
Ok, this comment had me worried for a minute, but ultimately these are not the same dictionaries.  What I see is that in `BaseTask.run`:

- `build_safe_env(env)` gets stored as `self.safe_env`
- `inject_credential` gets called for every credential, which updates `env` and `self.safe_env`
- `IsolatedManager.run()` dispatches this stuff including the unredacted contents of `env` out to the isolated node
- the unredacted env vars as seen on the isolated node are then serialized to the `config` artifact that I've added to ansible-runner
- `IsolatedManager.check()` captures the contents of this artifact and passes it back to `BaseTask.check_handler()`
- and then the `check_handler` callback does the redaction by calling `build_safe_env`

chrismeyersfsu replied:
Yep, so you just need to get a handle on self.safe_env to do the redaction, rather than calling build_safe_env in check_handler.

jbradberry replied:
I guess I'm just not following.  It seems like the artifact `env` will not necessarily fully match `self.safe_env` anymore by the time the isolated node is done with it, so it seems appropriate to sanitize fresh from the value in the artifact instead of copying the values in `self.safe_env`.  And it looks like the credentials injector uses this utility to do the sanitization in the first place, so this should be performing the exact same data manipulation with the same type of results when I use it in `check_handler`:

https://github.com/ansible/awx/blob/3f73176ef2ae0d952b02e579974c60b807ec172b/awx/main/models/credential/__init__.py#L684

Let's talk in person or on a call, though, to convince each other.

jbradberry replied:
Ok, I think this change will sufficiently address the situation.

jbradberry committed:
Fix a typo: there is no method called check_callback on BaseTask

jbradberry committed:
Bring the check_callback call out of the loop

We shouldn't need to call it multiple times.

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -191,6 +192,14 @@ def check(self, interval=None):
+            if os.path.exists(config_path):
+                with open(config_path, 'r') as f:
+                    data = json.load(f)
+                self.check_callback(data)
```
This still may be called multiple times, causing a database update each time it's called. We only need to update the database record once. Some sort of `self.called_check_callback_flag` here or in the callback registered outside would ensure that only 1 database update is called.

chrismeyersfsu replied:
This looks good.

chrismeyersfsu replied:
We want to save this invocation information as early as possible. Waiting for the job to finish is late. The information that we are saving is vital to debugging/recreating errors. Maybe check for the artifact on each loop and set a flag when you have processed that file and called the check_callback that was passed in? This way you get the benefit of calling it early, and don't incur extra database updates.

jbradberry replied:
Yeah, fair enough.  I'll come up with something.

jbradberry replied:
Got it, I think.

jbradberry committed:
Bring the check_callback back into the loop

but try to process it only once.

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -187,6 +189,19 @@ def check(self, interval=None):
+                            data = json.load(f)
+                        self.check_callback(data)
+                        self.captured_config_artifact = True
+                    except json.decoder.JSONDecodeError:  # Just in case it's not fully here yet.
```
nice

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -187,6 +189,19 @@ def check(self, interval=None):
+                        with open(config_path, 'r') as f:
+                            data = json.load(f)
+                        self.check_callback(data)
+                        self.captured_config_artifact = True
```
Good name, done after the json.load() :+1:

jbradberry committed:
Change the artifact file convention for isolated nodes to 'command'

since that's what landed in the ansible-runner PR.

jbradberry committed:
Capture the redacted credential env vars separately

and then make use of them specifically to make safe the env vars
coming back from an isolated node.  This will allow us to capture the
safed versions of custom credential values, but without potentially
clobbering normal env var values that vary between the controller and
the node.

ryanpetrello commented on awx/main/isolated/manager.py:
```diff
@@ -187,6 +189,19 @@ def check(self, interval=None):
+            if self.check_callback is not None and not self.captured_command_artifact:
+                command_path = self.path_to('artifacts', self.ident, 'command')
+                # If the configuration artifact has been synced back, update the model
+                if os.path.exists(command_path):
```
Nice work on this - this'll keep things working _until_ the new artifact lands in the next official runner.

ryanpetrello reviewed (approved):
I tested this locally by running a job against an isolated node.  Looks like it works to me, and properly filters sensitive env vars from the API:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/214912/55627586-8b87ba00-577c-11e9-9741-43255c843111.png">

It also picks up env vars set manually *on* the isolated node in `/etc/environment`:

<img width="1107" alt="image" src="https://user-images.githubusercontent.com/214912/55627734-e6b9ac80-577c-11e9-88fd-a82bb25f84d0.png">

kdelee reviewed (approved):
:shipit:
